### PR TITLE
Feature/gateway inbuilt nr

### LINKED ISSUE #3838: Introduce a local network requester directly inside a gateway

### PR DESCRIPTION
**Pull request by jstuczyn**
# Description

~~Depends on https://github.com/nymtech/nym/pull/3866 getting merged first~~

Closes: https://github.com/nymtech/nym/issues/3838

# Checklist:

- [ ] added a changelog entry to `CHANGELOG.md`

### CODE REVIEW

File: clients/native/src/client/config/mod.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ pub struct Config {
 }
 
 impl NymConfigTemplate for Config {
-    fn template() -> &'static str {
+    fn template(&self) -> &'static str {
```
Comment (octol) [inline clients/native/src/client/config/mod.rs:70]: Stylistic choice to add self here? Doesn't look like it's needed?

File: common/client-core/src/init/types.rs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    // pub fn is_custom_new(&self) -> bool {
+    //     if let GatewaySetup::New { specification } = self {
+    //         specification.is_custom()
+    //     } else {
+    //         false
+    //     }
+    // }
+
+    // pub async fn choose_gateway(
```
Comment (octol) [inline common/client-core/src/init/types.rs]: Commented out code?

File: gateway/src/commands/helpers.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    let Some(opts) = opts else { return cfg };
+
+    // as of 12.09.23 the below is true (not sure how this comment will rot in the future)
+    // medium_toggle:
+    // - sets secondary packet size to 16kb
+    // - disables poisson distribution of the main traffic stream
+    // - sets the cover traffic stream to 1 packet / 5s (on average)
+    // - disables per hop delay
+    //
+    // fastmode:
```
Comment (octol) [inline gateway/src/commands/helpers.rs]: side note for the future: `fastmode` should be renamed to `fast-poisson` or something

File: gateway/src/config/template.rs [hunk 4]
```diff
@@ -73,6 +77,9 @@ keys.public_sphinx_key_file = '{{ storage_paths.keys.public_sphinx_key_file }}'
 # derived shared keys and available client bandwidths.
 clients_storage = '{{ storage_paths.clients_storage }}'
 
+# Path to the configuration of the locally running network requester.
```
Comment (octol) [inline gateway/src/config/template.rs]: imho we should probably call it embedded or something, locally to me just means running on the same machine

File: gateway/src/node/mod.rs [hunk 5]
```diff
@@ -194,46 +194,114 @@ impl<St> Gateway<St> {
         packet_sender
     }
 
-    async fn wait_for_interrupt(
+    // TODO: rethink the logic in this function...
+    async fn maybe_start_network_requester(
```
Comment (octol) [inline gateway/src/node/mod.rs]: style: I prefer to have the check in the caller scope and not call it maybe ...

**Review by octol — APPROVED** [review]

File: clients/native/src/client/config/mod.rs (hunk 1 above)
Comment (jstuczyn) [reply clients/native/src/client/config/mod.rs:70]: I had an idea of providing different templates based on whether embedded NR is enabled or not. I think this still might be a useful feature in the further stages of project smoosh

**Pull request merged**